v2/client.go

```diff
@@ -1183,6 +1186,49 @@ func (c *client) FetchMetadata(ctx context.Context) (*Metadata, error) {
 	return c.metadata.update(ctx)
 }
 
+// FetchPartitionMetadata returns the metadata of the partition. This is sent to
+// the partition's leader
+func (c *client) FetchPartitionMetadata(ctx context.Context, stream string, partition int32) (*PartitionInfo, error) {
+	var (
+		req           = &proto.FetchPartitionMetadataRequest{Stream: stream, Partition: partition}
+		partitionInfo = &PartitionInfo{}
+	)
+
+	err := c.doResilientLeaderRPC(ctx, func(client proto.APIClient) error {
+		resp, err := client.FetchPartitionMetadata(ctx, req)
+		if err != nil {
+			return err
+		}
+		// Get broker info from metadata
+		brokers := c.metadata.metadata.brokers
```

Comment:
```suggestion
		brokers := c.metadata.get().Brokers()
```

Should use `get()` to avoid data races.

Reply:
For now `Brokers()` returns an array.

Should we just return the map and let the caller deal with it ?

```golang
// Brokers returns a list of the cluster nodes.
func (m *Metadata) Brokers() []*BrokerInfo {
	brokers := make([]*BrokerInfo, 0, len(m.brokers))
	for _, broker := range m.brokers {
		brokers = append(brokers, broker)
	}
	return brokers
}

```

For `Addrs()` it makes sense to return array, as most of the time use cases are to collect flat list of available addresses

```golang
// Addrs returns the list of known broker addresses.
func (m *Metadata) Addrs() []string {
	addrs := make([]string, 0, len(m.addrs))
	for addr := range m.addrs {
		addrs = append(addrs, addr)
	}
	return addrs
}
```

But in general, should we just return the map and let the caller deal with it in their own way ?  A map can be flattened to array but not vice-versa.

What do you think ?

Reply:
We can't change that API without requiring a new major version bump. Accessing the map internally is fine. The main thing is calling `get()` to avoid data races on the metadata object since this can get updated concurrently (the object itself is immutable however).